File: laba.py
import math
import matplotlib.pyplot as plt
import numpy as np

#инициализация всех входных данных
PROFILE=[150.2, 153.3, 171.6, 96, 125.1,155,158,134.5,134.5,112.7,83.1,102.2,151.4,109.5,148.4,83.2,180.1,180.1,191.8,177, 77.1,161.3, 203.3,118.3,171.5,203.1]
DLINA=[0,20,40,60,80,100,120,140,140,160,180,200,220,240,260,280,300,300,320,340,360,380,400,420,440,460]
DLINA[:] = [x * 1000 for x in DLINA]
NPS2 = 7
NPS3 = 16
n = 26
Pmax = 6.3*10**(6)
Do = 630/1000 # внешний 0.36371343685456636
b = 8/1000 #стенка
Di = Do-2*b
E = 0.2/1000/Di # шероховатость
plotnost = 850
viscosity = 15*10**(-6)
Py = 10*(10**3)
Hk = 35
Podpor = 40
Hmin = 40
Hmax = Pmax/(9.81*plotnost)
Head_max = [x+Hmax for x in PROFILE]
Hy = Py/(9.81*plotnost)
kavitacia = [0 for i in range(n)]
for i in range(len(PROFILE)-1):
    kavitacia[i] = Hy + PROFILE[i]
Head_x = [0 for i in range(n)]
Head_x[-1] = Hk+PROFILE[-1]
Head = [0 for i in range(n)]
Head[-1] = Hk+PROFILE[-1]
NUMBER_SAM = 0
FINAL_SAMOTEK = [[0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0]]

def Pumps(Q_FROM_PUMP, HEAD_FROM_PUMP): #МНК для насосов
    a = np.vstack([Q_FROM_PUMP**2, np.ones_like(Q_FROM_PUMP)]).T
    c = np.linalg.lstsq(a, HEAD_FROM_PUMP, rcond=None)[0]
    c[0] = c[0]*(-1)
    return c

Q_FROM_PUMP2 = np.array([1000, 1250, 1500, 1750, 2000]) # НПС-2 - НМ-4
HEAD_FROM_PUMP2 = np.array([260, 253, 245, 235, 222])
NASOS4 = Pumps(Q_FROM_PUMP2, HEAD_FROM_PUMP2) # коэф аппроксимации

Q_FROM_PUMP1 = np.array([1000, 1250, 1500, 1750, 2000]) # НПС-3 И ГНПС - НМ-2
HEAD_FROM_PUMP1 = np.array([243, 237, 233, 225, 218])
NASOS2 = Pumps(Q_FROM_PUMP1, HEAD_FROM_PUMP1) # коэф аппроксимации


def PUMPS_HEAD(Q, coeff_aprocs): # функция расчета дифф напора насосов
    diff_head = coeff_aprocs[1]-coeff_aprocs[0]*(Q*3600)**2
    return diff_head

def Soprot(Q): # считаю Lumbda
    Re = (Q * 4) / (math.pi*viscosity*Di)
    if Re <= 2320:
        gidroSoprot = 64 / Re
    if 2320 <= Re <= 10 * E**(-1):
        gidroSoprot = 0.3164 / (Re)**0.25
    if 10 * E**(-1) < Re < 500 * E**(-1):
        gidroSoprot = 0.11 *(E + (68 / Re))**0.25
    if Re >= 500 * E**(-1):
        gidroSoprot = 0.11 * E**0.25
    return gidroSoprot


def Head_loses (distance,Q):
    lymbda = Soprot(Q)
    lose_h = lymbda*distance*8*Q**2/(9.81*Di**5*math.pi**2)
    return lose_h


def delenie_popolam(foo, a, b, tol=0, max_iter=500):
    if foo(a) * foo(b) >= 0:
        print("Метод бисекции не применим")
        return None
    iter_count = 0
    while (b - a) / 2 > tol and iter_count < max_iter:
        c = (a + b) / 2
        if foo(c) == 0:
            return c
        elif foo(c) * foo(a) < 0:
            b = c
        else:
            a = c
        iter_count += 1
    return (a + b) / 2


def FI(F):
#нахождение fi для расчета самотека
    alpha = abs(math.atan(sam_angle))
    a1 = (F - math.sin(F)) ** (5 / 3) / F ** (2 / 3)
    a2 = 0.2419 * Q_last / (Di ** (8 / 3) * math.sqrt(math.sin(alpha)))
    otkl = a1 - a2
    return otkl


def Paschet(Q):
    NO_ITER=0
    #перезапись массивов двойников для каждого захода в foo()
    Head = [0 for i in range(n)]
    Head[-1] = Hk + PROFILE[-1]
    L = [0 for i in range(n)]
    for i in range(0, len(DLINA)):
        L[i] = DLINA[i]
    Z = [0 for i in range(n)]
    for i in range(0, len(DLINA)):
        Z[i] = PROFILE[i]
    Kav = [0 for i in range(n)]
    for i in range(0, len(DLINA)):
        Kav[i] = kavitacia[i]
    MAX = [0 for i in range(n)]
    for i in range(0, len(DLINA)):
        MAX[i] = Head_max[i]


    for i in range(n-2, -1, -1):
        Head[i] = Head[i + 1] + Head_loses(L[i + 1] - L[i], Q)

        if i==NPS2: # рассчет входа в станцию NPS2
            Head[i] = Head[i+1]-2*PUMPS_HEAD(Q, NASOS4)

        if i==NPS3: # рассчет входа в станцию
            Head[i] = Head[i+1]-2*PUMPS_HEAD(Q, NASOS2)

        # Проверка на отрицательное давление
        if Head[i]<0:
            NO_ITER=1

        if Head[i] < kavitacia[i] and NO_ITER==0: #ЕСЛИ НАШЛИ САМОТЕК ПРОВОДИМ ДОМ ПОСТРОЕНИЕ ТОЧКИ
            if (i == 0 or i == NPS2 or i == NPS3):
                print("Нарушено условие подпора")
            else:
                x_sam = L[i + 1] + (L[i + 1] - L[i]) * (Head[i + 1] - Z[i + 1] - Hy) / (
                            (Z[i + 1] - Z[i]) + (L[i + 1] - L[i]) * Head_loses(1, Q))

                h_sam = (Z[i + 1] - Z[i]) / (L[i + 1] - L[i]) * (x_sam - L[i]) + Z[i]

                L.insert(i + 1, x_sam)

                Z.insert(i + 1, h_sam)

                Head.insert(i + 1, Hy + Z[i + 1])

                MAX.insert(i + 1, Hmax + Z[i + 1])

                Kav.insert(i + 1, Hy + Z[i + 1])

                Head[i] = Hy + Z[i]

    a1=Head[0]

    a2=Podpor + 2*PUMPS_HEAD(Q, NASOS2)+ PROFILE[0]

    otkl=a2-a1

    print("несходимость метода деление отрезка пополам =", otkl)

    return otkl


# Поиск реального расхода с учетом самотечности
Q_last = delenie_popolam(Paschet, 0.2, 1.2)

for i in range(n - 2, -1, -1):
    Head[i] = Head[i + 1] + Head_loses(DLINA[i + 1] - DLINA[i], Q_last)

    if i == NPS2:
        Head[i] = Head[i + 1] - 2 * PUMPS_HEAD(Q_last, NASOS4)

    if i == NPS3:
        Head[i] = Head[i + 1] - 2 * PUMPS_HEAD(Q_last, NASOS2)

    # Проверка на отрицательное давление
    if Head[i] < 0:
        NO_ITER = 1

    if Head[i] < kavitacia[i]:
       NUMBER_SAM+=1
       x_sam = DLINA[i+1] + (DLINA[i + 1] - DLINA[i]) * (Head[i + 1] - PROFILE[i + 1] - Hy) / ((PROFILE[i + 1] - PROFILE[i])+(DLINA[i + 1] - DLINA[i])*Head_loses(1,Q_last))

       h_sam = (PROFILE[i + 1] - PROFILE[i]) / (DLINA[i + 1] - DLINA[i]) * (x_sam - DLINA[i]) + PROFILE[i]

       DLINA.insert(i + 1, x_sam)

       PROFILE.insert(i + 1, h_sam)

       Head.insert(i + 1, Hy + PROFILE[i + 1])

       Head_max.insert(i + 1, Hmax + PROFILE[i + 1])

       kavitacia.insert(i + 1, Hy + PROFILE[i + 1])

       Head[i] = Hy + PROFILE[i]

       sam_angle = (PROFILE[i] - h_sam)/(DLINA[i] - x_sam)

       F = delenie_popolam(FI,0.1, 2*math.pi)

       PART = (F - math.sin(F))/(2*math.pi)*100 # cтепень заполнения

       FINAL_SAMOTEK[NUMBER_SAM-1] = [DLINA[i],x_sam,PROFILE[i],h_sam, PART]



print("Расход [м3/c] =", Q_last)

if NUMBER_SAM > 0:
    print("Количество самотечных участков = ", NUMBER_SAM)

    print("(координата конца, координата начала, высотка конца, высотка начала, степень заполнения): 1 = ", FINAL_SAMOTEK[0][:])

    print("(координата конца, координата начала, высотка конца, высотка начала, степень заполнения): 2 = ", FINAL_SAMOTEK[1][:])

    print("(координата конца, координата начала, высотка конца, высотка начала, степень заполнения): 3 = ", FINAL_SAMOTEK[2][:])

    print("(координата конца, координата начала, высотка конца, высотка начала, степень заполнения): 4 = ", FINAL_SAMOTEK[3][:])

    print("(координата конца, координата начала, высотка конца, высотка начала, степень заполнения): 5 = ", FINAL_SAMOTEK[4][:])


if NUMBER_SAM == 0:
    print("Самотечные участки отсутствуют")

# График ГУ
fig = plt.figure()
ax = fig.add_subplot()
ax.set_xlabel('Расстояние, км')
ax.set_ylabel('Напор, м')

plt.plot(DLINA, PROFILE, label = "Высотные отметки Z(X)")
plt.plot(DLINA, Head_max, label = "Несущая способность Hmax(X)")
plt.plot(DLINA, Head, label = "Напор H(X)")

plt.grid()
plt.legend()

# график зависимости давления от координаты
Pmax_mass = [Pmax/1000000 for i in range(len(Head))]
Pressure = [0 for i in range(len(Head))]
for i in range(len(Head)-1, -1, -1):
    Pressure[i] = (Head[i]-PROFILE[i])*(plotnost*9.81)/1000000
fig = plt.figure()
ax = fig.add_subplot()
ax.set_xlabel('Расстояние, км')
ax.set_ylabel('Давление, МПа')

plt.plot(DLINA, Pressure, label = "Давление P(X)")
plt.plot(DLINA, Pmax_mass, label = "Несущая способность Pmax(X)")

plt.grid()
plt.legend()
plt.show()

File: laba.cpp
#include <Eigen/Dense>
#include <cmath>
#include <iostream>
#include <vector>

// TODO: uppercase globals
std::vector<double> PROFILE = {150.2, 153.3, 171.6, 96,    125.1, 155,   158,
                               134.5, 134.5, 112.7, 83.1,  102.2, 151.4, 109.5,
                               148.4, 83.2,  180.1, 180.1, 191.8, 177,   77.1,
                               161.3, 203.3, 118.3, 171.5, 203.1};
std::vector<double> DLINA = {
    0,      20000,  40000,  60000,  80000,  100000, 120000, 140000, 140000,
    160000, 180000, 200000, 220000, 240000, 260000, 280000, 300000, 300000,
    320000, 340000, 360000, 380000, 400000, 420000, 440000, 460000};
double b = 8.0 / 1000.0;
double Do = 630.0 / 1000.0;
double Di = Do - 2 * b;
double E = 0.2 / 1000 / Di;
double viscosity = 15 * pow(10, (-6));
int NPS2 = 7;
int NPS3 = 16;
int n = 26;
double Pmax = 6.3 * pow(10, 6);
double plotnost = 850;
double Py = 10 * pow(10, 3);
double Hk = 35;
double Podpor = 40;
double Hmin = 40;
double Hmax = Pmax / (9.81 * plotnost);
double Q_last = 0;
int NUMBER_SAM = 0;
double sam_angle = 0;
double Hy = Py / (9.81 * plotnost);
double x_sam = 0;
double h_sam = 0;
std::vector<double> Pmax_mass;
std::vector<double> Pressure;
std::vector<double> Head_max(PROFILE);
std::vector<double> kavitacia(n, 0);
std::vector<double> Head_x(n, 0);
std::vector<double> Head(n, 0);
std::vector<double> Q_FROM_PUMP2{1000, 1250, 1500, 1750, 2000};
std::vector<double> Q_FROM_PUMP1{1000, 1250, 1500, 1750, 2000};
std::vector<double> HEAD_FROM_PUMP2{260, 253, 245, 235, 222};
std::vector<double> HEAD_FROM_PUMP1{243, 237, 233, 225, 218};
std::vector<double> NASOS2;
std::vector<double> NASOS4;
std::vector<std::vector<double>> FINAL_SAMOTEK{{0, 0, 0, 0, 0},
                                               {0, 0, 0, 0, 0},
                                               {0, 0, 0, 0, 0},
                                               {0, 0, 0, 0, 0},
                                               {0, 0, 0, 0, 0}};

void CountGlobals();
void CalculateConsumption();
double CalculateDeviation(double);
void CalculateGravityFlowAreas();
void CalculatePressure();
void PlotEverything();
double Bisection(double, double, double (*)(double));
std::vector<double> SolveLlss(std::vector<double>, std::vector<double>);

int main() {
  CountGlobals();
  CalculateConsumption();
  CalculateGravityFlowAreas();
  CalculatePressure();
  PlotEverything();
  return 0;
}

void CountGlobals() {
  for (int i = 0; i < n; i++) {
    Head_max[i] += Hmax;
    kavitacia[i] = Hy + PROFILE[i];
  }
  kavitacia[n - 1] = 0;
  Head_x[n - 1] = Hk + PROFILE[n - 1];
  Head[n - 1] = Hk + PROFILE[n - 1];
  NASOS2 = SolveLlss(Q_FROM_PUMP1, HEAD_FROM_PUMP1);
  NASOS4 = SolveLlss(Q_FROM_PUMP2, HEAD_FROM_PUMP2);
}

std::vector<double> SolveLlss(std::vector<double> Q_FROM_PUMP,
                              std::vector<double> HEAD_FROM_PUMP) {
  std::vector<double> res;
  Eigen::Matrix<double, 5, 2> A;
  Eigen::Matrix<double, 5, 1> b;
  for (int i = 0; i < 5; i++) {
    A.coeffRef(i, 0) = pow(Q_FROM_PUMP[i], 2);
    A.coeffRef(i, 1) = 1;
    b[i] = HEAD_FROM_PUMP[i];
  }
  auto Res = A.colPivHouseholderQr().solve(b);
  res.resize(Res.size());
  for (int i = 0; i < Res.size(); i++)
    res[i] = Res[i];
  res[0] *= (-1);
  return res;
}
// TODO: plots
void PlotEverything() { std::cout << "Here will be plots" << std::endl; }

double Bisection(double a, double b, double (*function)(double)) {
  if (function(a) * function(b) >= 0) {
    std::cout << "Метод бисекции не применим" << std::endl;
    return 0.0;
  }
  int iter_count = 0;
  while ((b - a) / 2 > 0 && iter_count < 500) {
    double c = (a + b) / 2;
    if (function(c) == 0)
      return c;
    else if (function(c) * function(a) < 0)
      b = c;
    else
      a = c;
    iter_count++;
  }
  return (a + b) / 2;
}

double Resistance(double Q) {
  double Re = (Q * 4) / (M_PI * viscosity * Di);
  double hydraulicResistance = 0;
  if (Re <= 2320)
    hydraulicResistance = 64 / Re;
  if (2320 <= Re <= 10 * pow(E, (-1)))
    hydraulicResistance = 0.3164 / pow(Re, 0.25);
  if (10 * pow(E, (-1)) < Re < 500 * pow(E, (-1)))
    hydraulicResistance = 0.11 * pow((E + (68 / Re)), 0.25);
  if (Re >= 500 * pow(E, (-1)))
    hydraulicResistance = 0.11 * pow(E, 0.25);
  return hydraulicResistance;
}

double Fi(double f) {
  double alpha = fabs(atan(sam_angle));
  double a1 = pow((f - sin(f)), (5.0 / 3.0)) / pow(f, (2.0 / 3.0));
  double a2 = 0.2419 * Q_last / (pow(Di, (8.0 / 3.0)) * sqrt(sin(alpha)));
  double deviation = a1 - a2;
  return deviation;
}

double PumpsHead(double Q, std::vector<double> coeff_aprocs) {
  double diff_head = coeff_aprocs[1] - coeff_aprocs[0] * pow((Q * 3600), 2);
  return diff_head;
}

double HeadLoses(double distance, double Q) {
  double lymbda = Resistance(Q);
  double lose_h =
      lymbda * distance * 8 * pow(Q, 2) / (9.81 * pow(Di, 5) * pow(M_PI, 2));
  return lose_h;
}

double CalculateDeviation(double Q) {
  int no_iter = 0;
  Head = std::vector<double>(n, 0);
  Head[n - 1] = Hk + PROFILE[n - 1];
  std::vector<double> L(n, 0);
  for (int i = 0; i < n; i++)
    L[i] = DLINA[i];
  std::vector<double> Z(n, 0);
  for (int i = 0; i < n; i++)
    Z[i] = PROFILE[i];
  std::vector<double> Kav = kavitacia;
  std::vector<double> MAX = Head_max;
  for (int i = n - 2; i >= 0; i--) {
    Head[i] = Head[i + 1] + HeadLoses(L[i + 1] - L[i], Q);
    if (i == NPS2)
      Head[i] = Head[i + 1] - 2 * PumpsHead(Q, NASOS4);
    if (i == NPS3)
      Head[i] = Head[i + 1] - 2 * PumpsHead(Q, NASOS2);
    if (Head[i] < 0)
      no_iter = 1;
    if (Head[i] < kavitacia[i] && no_iter == 0) {
      if (i == 0 || i == NPS2 || i == NPS3) {
        std::cout << "Нарушено условие подпора" << std::endl;
      } else {
        x_sam = L[i + 1] +
                (L[i + 1] - L[i]) * (Head[i + 1] - Z[i + 1] - Hy) /
                    ((Z[i + 1] - Z[i]) + (L[i + 1] - L[i]) * HeadLoses(1, Q));
        h_sam = (Z[i + 1] - Z[i]) / (L[i + 1] - L[i]) * (x_sam - L[i]) + Z[i];
        L.insert(L.begin() + (i + 1), x_sam);
        Z.insert(Z.begin() + (i + 1), h_sam);
        Head.insert(Head.begin() + (i + 1), Hy + Z[i + 1]);
        MAX.insert(MAX.begin() + (i + 1), Hmax + Z[i + 1]);
        Kav.insert(Kav.begin() + (i + 1), Hy + Z[i + 1]);
        Head[i] = Hy + Z[i];
      }
    }
  }
  double a1 = Head[0];
  double a2 = Podpor + 2 * PumpsHead(Q, NASOS2) + PROFILE[0];
  return a2 - a1;
}

void CalculateConsumption() {
  Q_last = Bisection(0.2, 1.2, &CalculateDeviation);
  std::cout << "Расход [м3/c] = " << Q_last << std::endl;
}

void CalculateGravityFlowAreas() {
  for (int i = n - 2; i >= 0; i--) {
    Head[i] = Head[i + 1] + HeadLoses(DLINA[i + 1] - DLINA[i], Q_last);
    if (i == NPS2)
      Head[i] = Head[i + 1] - 2 * PumpsHead(Q_last, NASOS4);
    if (i == NPS3)
      Head[i] = Head[i + 1] - 2 * PumpsHead(Q_last, NASOS2);
    if (Head[i] < kavitacia[i]) {
      NUMBER_SAM++;
      x_sam = DLINA[i + 1] +
              (DLINA[i + 1] - DLINA[i]) * (Head[i + 1] - PROFILE[i + 1] - Hy) /
                  ((PROFILE[i + 1] - PROFILE[i]) +
                   (DLINA[i + 1] - DLINA[i]) * HeadLoses(1, Q_last));

      h_sam = (PROFILE[i + 1] - PROFILE[i]) / (DLINA[i + 1] - DLINA[i]) *
                  (x_sam - DLINA[i]) +
              PROFILE[i];
      DLINA.insert(DLINA.begin() + (i + 1), x_sam);
      PROFILE.insert(PROFILE.begin() + (i + 1), h_sam);
      Head.insert(Head.begin() + (i + 1), Hy + PROFILE[i + 1]);
      Head_max.insert(Head_max.begin() + (i + 1), Hmax + PROFILE[i + 1]);
      kavitacia.insert(kavitacia.begin() + (i + 1), Hy + PROFILE[i + 1]);
      Head[i] = Hy + PROFILE[i];
      sam_angle = (PROFILE[i] - h_sam) / (DLINA[i] - x_sam);
      double f = Bisection(0.1, 2 * M_PI, &Fi);
      double part = ((f - sin(f)) / (2 * M_PI) * 100);
      FINAL_SAMOTEK[NUMBER_SAM - 1] = {DLINA[i], x_sam, PROFILE[i], h_sam,
                                       part};
    }
  }
  if (NUMBER_SAM > 0) {
    std::cout << "Количество самотечных участков = " << NUMBER_SAM << std::endl;
    for (int i = 0; i < 5; i++) {
      std::cout << "(координата конца, координата начала, высотка конца, "
                   "высотка начала, степень заполнения): "
                << i + 1 << " = " << FINAL_SAMOTEK[i][0] << FINAL_SAMOTEK[i][1]
                << FINAL_SAMOTEK[i][2] << FINAL_SAMOTEK[i][3]
                << FINAL_SAMOTEK[i][4] << std::endl;
    }
  } else {
    std::cout << "Самотечные участки отсутствуют" << std::endl;
  }
}

void CalculatePressure() {
  Pmax_mass = std::vector<double>(Head.size(), Pmax / 1000000);
  Pressure = std::vector<double>(Head.size(), 0);
  for (int i = Head.size() - 1; i >= 0; i--)
    Pressure[i] = (Head[i] - PROFILE[i]) * (plotnost * 9.81) / 1000000;
}
